crates/cranelift/src/wasmfx/optimized.rs

```diff
@@ -2191,3 +2311,294 @@ pub(crate) fn translate_suspend<'a>(
+
+        (switchee_contref_csi, switchee_contref_last_ancestor)
+    };
+
+    // Update VMContext/Store: Update active continuation and `VMRuntimeLimits`.
+    {
+        vmctx.set_active_continuation(env, builder, switchee_contref.address);
+
+        switchee_contref_csi.write_limits_to_vmcontext(env, builder, vm_runtime_limits_ptr);
+    }
+
+    // Perform actual stack switch
+    {
+        let switcher_last_ancestor_fs =
+            switcher_contref_last_ancestor.get_fiber_stack(env, builder);
+        let switcher_last_ancestor_cc =
+            switcher_last_ancestor_fs.load_control_context(env, builder);
+
+        let switchee_last_ancestor_fs =
+            switchee_contref_last_ancestor.get_fiber_stack(env, builder);
+        let switchee_last_ancestor_cc =
+            switchee_last_ancestor_fs.load_control_context(env, builder);
+
+        // The stack switch involves the following control contexts (e.g., IP,
+        // SP, FP, ...):
+        // - `switchee_last_ancestor_cc` contains the information to continue
+        //    execution in the switchee/target continuation.
+        // - `switcher_last_ancestor_cc` contains the information about how to
+        //    continue execution once we suspend/return to the stack with the
+        //    switch handler.
+        //
+        // In total, the following needs to happen:
+        // 1. Load control context at `switchee_last_ancestor_cc` to perform
+        //    stack switch.
+        // 2. Move control context at `switcher_last_ancestor_cc` over to
+        //    `switchee_last_ancestor_cc`.
+        // 3. Upon actual switch, save current control context at
+        //    `switcher_last_ancestor_cc`.
+        //
+        // We implement this as follows:
+        // 1. We copy `switchee_last_ancestor_cc` to a temporary area on the
+        //    stack (`tmp_control_context`).
+        // 2. We copy `switcher_last_ancestor_cc` over to
+        //    `switchee_last_ancestor_cc`.
+        // 3. We invoke the stack switch instruction such that it reads from the
+        //    temporary area, and writes to `switcher_last_ancestor_cc`.
+        //
+        // Note that the temporary area is only accessed once by the
+        // `stack_switch` instruction emitted later in this block, meaning that we
+        // don't have to worry about its lifetime.
+        //
+        // NOTE(frank-emrich) We could avoid the copying to a temporary area by
+        // making `stack_switch` do all of the necessary moving itself. However,
+        // that would be a rather ad-hoc change to how the instruction uses the
+        // two pointers given to it.
+
+        let slot_size = ir::StackSlotData::new(
+            ir::StackSlotKind::ExplicitSlot,
+            CONTROL_CONTEXT_SIZE as u32,
+            env.pointer_type().bytes() as u8,
+        );
+        let slot = builder.create_sized_stack_slot(slot_size);
+        let tmp_control_context = builder.ins().stack_addr(env.pointer_type(), slot, 0);
```

Comment:
We have no data to validate or refute the assumption. I suggest we add a comment to remark that there may be an issue here, and if it turns out to be issue then there is a straightforward cache-fix by extending the `FuncEnvironment`.

Reply:
It would be interesting to experiment with using stack slots to encode "handlerlists". I don't know whether it would offer any speedups, but it would eliminate some heap memory dependency. Similarly, it ought to be possible to write resume/suspend/switch arguments directly on the target stack using the stack slots API, thus potentially eliminating the need for variable-sized vectors.

Reply:
I've added a comment now about the possibility of re-using the stack slots.

> It would be interesting to experiment with using stack slots to encode "handlerlists". 

Yes, I thought about that a while ago. It should be relatively straightforward, with one annoyance: For the initial payloads buffer (i.e., the one consumed by the array call trampoline, represented by the `args` field in `VMContRef`), we need to do the reservation of some stack space ourselves, somewhere at the top of the fiber stack. We cannot use the ordinary stack slot API for that. But this is mostly a matter of updating some ASCII art and tweaking the initialization functions.